File: frontend/src/app/data/product.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { ErrorHandler, Injectable } from '@angular/core';
import { ProductFilter } from '../domain/entities/product.filter.entity';
import { catchError, Observable } from 'rxjs';
import { Product } from '../models/product.model';
import { ProductRepositoryImpl } from '../domain/repository/product.repository.impl';

@Injectable({
  providedIn: 'root'
})
// Focus on connecting repository that retrieves raw datra from backend and convert to application model data
//, apply business logic and displayy to view and handle user interaction.
export class ProductService {

  constructor(private repository: ProductRepositoryImpl) { }

  getProducts(page: number, pageSize: number): Observable<Product[]> {
    return this.repository.getAll(page, pageSize);
  }

  getProductsById(id: number): Observable<Product> {
    return this.repository.getById(id);
  }

  getProductsByType(type: string): Observable<Product> {
    return this.repository.getByType(type);
  }

  getProductsByPrice(price: number): Observable<Product> {
    return this.repository.getByPrice(price);
  }

  deleteProduct(id: number): Observable<void> {
    return this.repository.delete(id);
  }

  searchProducts(name: string): Observable<Product[]> {
    return this.repository.searchByName(name);
  }

  createProduct(product: Product): Observable<Product> {
    return this.repository.create(product);
  }

  updateProduct(id: number, product: Product): Observable<Product> {
    return this.repository.update(id, product);
  }
}

File: app/src/main/java/com/example/product_list_app/services/ProductService.kt
package com.example.product_list_app.services

import com.example.product_list_app.model.ProductData
import com.example.product_list_app.model.ProductList
import com.example.product_list_app.repository.ProductDataRepositoryImpl

class ProductService(private val repo: ProductDataRepositoryImpl ) {
    suspend fun getProducts(page: Int, limit: Int): Result<ProductList> {
        return repo.getProducts(page, limit)
    }

    suspend fun getProductById(id: Int): Result<ProductData> {
        return repo.getProductById(id)
    }

    suspend fun searchProducts(query: String): Result<List<ProductData>> {
        return repo.searchProducts(query)
    }

    suspend fun createProduct(product: ProductData): Result<ProductData> {
        return repo.createProduct(product)
    }

    suspend fun updateProduct(id: Int, product: ProductData): Result<ProductData> {
        return repo.updateProduct(id, product)
    }

    suspend fun deleteProduct(id: Int): Result<Void> {
        return repo.deleteProduct(id)
    }
}